tools/packages/wesl/src/VirtualFilesystem.ts

```diff
@@ -0,0 +1,89 @@
+import { assertThat } from "./Assertions";
+import { isIdent } from "./parse/WeslStream";
+
+/**
+ * A relative, normalized, Linux-style path ending with a file.
+ * All file and folder names must be valid WGSL identifiers.
+ * `.wgsl` and `.wesl` are the supported file extensions.
+ */
+export type WeslPath = string & { __weslPath: never };
+
+/** A async virtual filesystem, can be backed by an in-memory map, or by a real filesystem, or by HTTP requests */
+export interface VirtualFilesystem {
+  /**
+   * The WESL module loading uses this to load files.
+   */
+  readFile(path: WeslPath): Promise<string | null>;
+}
+
+/** Creates a static filesystem from relative, normalized, Linux-style paths.
+ * All file and folder names must be valid WGSL identifiers.
+ * `.wgsl` and `.wesl` are the supported file extensions.
+ */
+export function staticFilesystem(
+  files: Record<string, string>,
+): VirtualFilesystem {
+  // Avoid accidentally having a file called `toString`
+  const filesystemMap = new Map(
+    Object.entries(files).map(([path, contents]) => [
+      makeWeslPath(path),
+      contents,
+    ]),
+  );
+  function readFile(path: WeslPath): Promise<string | null> {
```

Comment:
Notice how I'm assuming that the path is well behaved, instead of doing normalization.
This is for two reasons
1. This function only gets called by the linker internals, where the paths are well behaved. So doing normalization is pointless.
2. Normalization shouldn't be done in this case. Instead the filesystem implementation should directly query the filesystem with a given path. This lets us correctly deal with symlinks.